### ISSUE TITLE
Problem with max pool size was reached.

### ISSUE THREAD

**Issue by vzaoral** [dotnet/efcore#6603]
### Steps to reproduce

Hello. We're using EF Core for querying our DB in our WebApi app. Sometimes when we have big traffic, querying to DB ends with this error:

> Timeout expired. The timeout period elapsed prior to obtaining a connection from the pool. This may have occurred because all pooled connections were in use and max pool size was reached.

I wonder if our pattern of using dbcontext and querying is correct or if I am missing some _using_/dispose pattern and error is caused by some memory leak (after some research I read then I should not use _using_ because the lifetime is managed by the framework). I am following [documentation](https://docs.efproject.net/en/latest/miscellaneous/configuring-dbcontext.html#using-dbcontext-with-dependency-injection)...
### The issue

my connectionString
`"myConnection": "Server=xxx;Database=xxx;user id=xxx;password=xxx;Max Pool Size=200;Timeout=200;"`

My Startup.cs

```
public void ConfigureServices(IServiceCollection services)
        {
        .....
        // scoped context            
        services.AddDbContext<MyDbContext>(
                options => options.UseSqlServer(this.Configuration.GetConnectionString("myConnection")));
        }
```

then in controllers I used dbcontext by dependency injection:

```
public class MyController : Controller
       public MyController (MyDbContext context)
        {
            this.Context = context;            
        }

        public ActionResult Get(int id)
        {
            // querying
            return this.Context.tRealty.Where(x=>x.id == id).FirstOrDefault();
        }
```
### Further technical details

EF Core version: 1.0.1
Operating system: WIN10
Visual Studio version: VS2015 Update 3

**Comment by divega** [comment]
@rowanmiller this looks very similar to another potential "stress" issue assign to you. Could you take a look?

**Comment by vzaoral** [comment]
So, now is our api running without _max pool size was reached_ errors. I don't know if there is some connection, but we were using [In Memory Caching](https://docs.asp.net/en/latest/performance/caching/memory.html) in our app to cache results from database queries. Now we are using [Redis Distributed Cache](https://docs.asp.net/en/latest/performance/caching/distributed.html#using-a-redis-distributed-cache) and everything is working fine. So hard to say...

**Comment by vzaoral** [comment]
I think problem was caused by storing objects from database context queries to _In memory cache_. I had one big LINQ query to database context with some other subqueries inside. I called `FirstOrDefault()` on the end of main query **but not inside subqueries**. Controller was fine with it, it materialize queries by default.


```
return this.Context.tRealty.AsNoTracking().Where(
                    x => x.Id == id && x.RealtyProcess == RealtyProcess.Visible).Select(
                    s => new
                    { .....

    // subquery
    videos = s.TVideo.Where(video => video.RealtyId == id && video.IsPublicOnYouTube).
                            Select(video => video.YouTubeId).ToList()), // missing ToList()
    .....
     }).FirstOrDefault();
```


And there was problem - subqueries were holding connection to database context when they where storing to _In memory cache_. When I implemented _Redis distributed cache_, it was first failing on some strange errors. It helps when I write `ToList()`or `FirstOrDefault()` to all my subqueries because distributed cache needs materialized objects. 

Now I have all my queries materialized explicitly and I got no _max pool size was reached error_. So that one must be careful when stored objects from database context queries to _In memory cache_. It is need to materialize all queries to avoid to holding connection somewhere in memory.
